### PR TITLE
[ts] Handle nullable layer members in layer code

### PR DESCRIPTION
**Pull request by Pessimistress**
For #5589 

This proposal replaces many `@ts-ignore` usage with the `!` pattern. I think it is a better approach because:
- `@ts-ignore` ignores ALL TS rules, not just 2531. The decorated line practically loses all type checks, which could be valuable in catching other bugs. `!` only tells TS to ignore one variable's possible null state.
- It's just less verbose to write.

I purposefully do not use this pattern in the core module - where things are more subject to unexpected states - I believe that it's better to explicitly document any assumptions there. For the layer implementations, however, the lifecycle methods are only ever invoked by the layer manager, where `state`, `context` and `internalState` are always defined. The risk that anything will break by forcing TypeScript to ignore the nullable members are extremely low.

### CODE REVIEW

File: modules/layers/src/scatterplot-layer/scatterplot-layer.ts [hunk 1]
```diff
@@ -103,8 +103,7 @@ export default class ScatterplotLayer<DataT = any, ExtraPropsT = {}> extends Lay
   }
 
   initializeState() {
-    // @ts-ignore (TS2531) attributeManager is always defined for primitive layer
-    this.getAttributeManager().addInstanced({
+    this.getAttributeManager()!.addInstanced({
```
Comment (ibgreen) [inline modules/layers/src/scatterplot-layer/scatterplot-layer.ts:106]: Should we really allow AttributeManager to be undefined? Can't we prototype getAttributeManager to always return  a valid instance, and do checks during initialization instead?

**Review by ibgreen — APPROVED** [review]
>  the ! pattern. I think it is a better approach because:

In general am not a fan of the exclamation mark. It is very subtle in the code, hard to spot, and can end up hiding very-difficult-to-detect bugs. I would gently vote against it, especially as a widespread pattern, and try hard to fix the various situations that cause us to need it.

One observation I can offer is that I essentially never use `ts-ignore-error` which is part of your argument (and I agree it is very bad). 

I always prefer `ts-expect-error`, which errors out once the underlying issue is fixed.

In most cases, `ts-ignores` are needed because other parts of the code haven't yet been fully typed. Having `ts-expect-error` makes it easy to find and remove exceptions as code improves, and I have been removing these at a good clip in other frameworks.

**Comment by Pessimistress** [comment]
@ibgreen I'm open to suggestions on how to fix the underlying issue in this case. A layer's state, context and internal state are null until it is initialized or matched. We cannot type these members as always present because then we lose important typecheck in the base layer class and the layer manager. But they are almost never null when you implement a layer's lifecycle methods.

**Comment by ibgreen** [comment]
I see. Just brainstorming, 

Maybe the can be accessed via protected members in the base classes, which do expose the conditionality, and via public `getAttributeManager()` style helper methods in the sub classes.

Even if that isn't perfect, it would optimize somewhat for one of the two cases (internals vs subclasses). Making it easier for subclass developers (while asking internal code to avoid `getAttributeManager()`) seems like it could be a reasonable choice.

**Comment by Pessimistress** [comment]
An alternative is to use a getter for state/context/internalState, along the lines of: 

```js
_state: Record<string, any> | null = null;

set state(value: Record<string, any> {
  // State should only be assigned once when the layer is initialized/matched
  if (this._state === null) {
    this._state = value;
  }
}

get state: Record<string, any> {
  assert(this._state, 'Layer does not have state');
  return this._state;
}
```

I'm a little concerned about the perf implications (especially if state is accessed during attribute update such as https://github.com/visgl/deck.gl/blob/8.7-release/modules/layers/src/icon-layer/icon-layer.js#L261)

**Comment by ibgreen** [comment]
> I'm a little concerned about the perf implications

Yes... a quick round of profiling on 2-3 browsers could perhaps give us a sense. 

But unless we are concerns about modifying state maybe we could just do with:

```
protected _state: LayerState | null = null;
state: LayerState;
```

I think having a well commented ts-expect-error in one place is OK. Just make the sure the commented line is minimal (break out the problematic line from any surrounding expression), then we can just do:
```
protected _state: LayerState | null = null;
readonly state: LayerState;
```

**Comment by Pessimistress** [comment]
I do not see a big difference in Chrome, but FF and Safari see significant overhead (> 2x) when using a getter. Even without assertion the perf impact is there.

**Comment by zbigg** [comment]
Back when i started i've got fee ideas, none of them satisfactory, but i'll write them down:

1) getters, and `!` already discussed
2) assert on `this.state` everywhere - also already discussed
3) type assertion on `this`, random examples:

```
interface ProperLayer<LayerT> {
     state: NonNullable<Layer["state"]);
     // same for context & internal state
}

class LayerX {
  foo() {
      (this as ProperLayer<this>).updateState() // ok
      (this as ProperLayer<this>).updateState() // ok
  }
  updateStateInt(this: ProperLayer<this>, ...) {
      this.updateStateInt() // ok
  }

  updateStateInt(this: ProperLayer<this>, ...) {
      const { fetch } = this.state ; // ok
  }
  
  orANotherApproachWhicKindaUgly() {
      const safeThis = this as ProperLayer<this>
      const { fetch } = this.state ; // ok
  }
}
```

4. define `state` as mandatory and _indeed_ initialize it in Layer but with magic value that is treated as null

```
const DEFAULT_STATE_NOT_INITIALIZED = Object.freeze({}) 
// freezed only to ensure nobody will be able to write anything there
const 
Component {
    state: Record<string, any> = DEFAULT_STATE_NOT_INITIALIZED;
}
```
but that will probably fail with compatibility with external code, which for sure used `null` as indicator of non-initialized Layer

**Comment by Pessimistress** [comment]
I reviewed the usage of `context` and `state` in the base Layer class. They are overwhelmingly used in these conditions:

- internal lifecycle methods
- preceded by an assertion of `internalState`

I have changed their declaration to always present. This removes the need to test for null state in layer implementations.

`getAttributeManager()` still may return `null`.
